### DISCUSSION TITLE
Strange Timing in SPI

### DISCUSSION THREAD

**Discussion by RudolfAtHome** [earlephilhower/arduino-pico#330] [Q&A] [answered]
I found a strange timing in SPI.cpp and analyzed it:
The "beginTransaction (...)" function takes 67 µs! The following 16-bit SPI transfer then takes 1.09 µs. Why is SPI slowed down so much?
I analyzed the function calls and probably identified the cause:

  **//---- in \rp2040\hardware\rp2040\1.9.4\libraries\SPI ----**_
  __void SPIClassRP2040::beginTransaction(SPISettings settings)
      utBegin2 = micros();                       // My time measure start
      spi_init(_spi, _spis.getClockFreq());  // -> the blocking function
      utBegin2 = micros() - utBegin2;	**// My time measure end -> 67 µs**__
	//---- in \rp2040\hardware\rp2040\1.9.4\pico-sdk\src\rp2_common\hardware_spi\spi.c ----
	uint spi_init(spi_inst_t *spi, uint baudrate)
          static inline void spi_unreset(spi_inst_t *spi)
          unreset_block_wait(spi == spi0 ? RESETS_RESET_SPI0_BITS : RESETS_RESET_SPI1_BITS);
          static inline void unreset_block_wait(uint32_t bits)	**// -> this is the cause (?)**_

**Comment by earlephilhower** [comment, marked as answer]
Interesting observation!  I imagine the clock startup time could be pretty long.

Maybe we can only call `spi_init()` when the SPI clock frequency changes, and never disable the SPI block until a `SPI.end()`!  It feels like a simple tweak, do you think you could give it a try?  Basically, cache the last clock and options and if they're the same as before we might be able to skip the init.  OTW, call the spi deinit/spi reinit and cache these new settings...

**Reply by earlephilhower** [reply]
See #394 which should fix this and speed up transactions a lot.
